### PR TITLE
COO intersection primitives : fusing value selection with value intersection.

### PR DESCRIPTION
**Pull request by nikitaved**
As per title. This one fuses 3 kernels into 1 with about 20-10% performance improvement.
This kernel is also useful for union-like operations.

### CODE REVIEW

File: aten/src/ATen/native/sparse/SparseBinaryOpIntersectionKernel.cpp [hunk 1]
```diff
@@ -15,18 +16,82 @@ struct CPUKernelLauncher {
...
+                  const auto* ptr_rhs_values_bytes = data[3];
+                  const auto* ptr_rhs_select_idx_bytes = data[4];
+
+                  for (int64_t i = 0; i < n; ++i) {
+                    // Exctract data
+                    auto* RESTRICT ptr_res_values = reinterpret_cast<scalar_t*>(ptr_res_values_bytes);
+                    const auto* ptr_lhs_values = reinterpret_cast<const scalar_t*>(ptr_lhs_values_bytes);
+                    const auto lhs_nnz_idx = *reinterpret_cast<const index_t*>(ptr_lhs_select_idx_bytes);
+                    const auto* ptr_rhs_values = reinterpret_cast<const scalar_t*>(ptr_rhs_values_bytes);
+                    const auto rhs_nnz_idx = *reinterpret_cast<const index_t*>(ptr_rhs_select_idx_bytes);
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/SparseBinaryOpIntersectionKernel.cpp:66]: @amjames , could you please confirm whether skipping `restrict` for these value pointers is justified?

File: aten/src/ATen/native/cuda/SparseBinaryOpIntersectionKernel.cu [hunk 2]
```diff
@@ -16,16 +18,127 @@ struct CUDAKernelLauncher {
...
+
+  auto offset_calc = make_offset_calculator<5>(iter);
+  auto loop = [=] FUNCAPI (int i) {
+    auto offsets = offset_calc.get(i);
+
+    auto* RESTRICT ptr_res_values = reinterpret_cast<scalar_t*>(ptr_res_values_bytes + offsets[0]);
+    const auto* ptr_lhs_values = reinterpret_cast<const scalar_t*>(ptr_lhs_values_bytes + offsets[1]);
+    const auto lhs_nnz_idx = *reinterpret_cast<const index_t*>(ptr_lhs_select_idx_bytes + offsets[2]);
+    const auto* ptr_rhs_values = reinterpret_cast<const scalar_t*>(ptr_rhs_values_bytes + offsets[3]);
+    const auto rhs_nnz_idx = *reinterpret_cast<const index_t*>(ptr_rhs_select_idx_bytes + offsets[4]);
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/SparseBinaryOpIntersectionKernel.cu:88]: @amjames , could you please confirm whether skipping restrict for these value pointers is justified?

Comment (amjames) [reply aten/src/ATen/native/cuda/SparseBinaryOpIntersectionKernel.cu:88]: This is a bit of an odd case, you have assigned RESTRICT to the output pointer. The result used for the TensorIterator is always allocated? If the out variant is called the values are copied to the provided tensor after the fact? 

If we already know that the only write address does not overlap then it should be safe for the compiler to load all read only values once, and writes to the output should not invalidate cached values.

So either restrict should be omitted on the result, or if we can be sure of that address' exclusivity then all pointer may safely be restricted.

Comment (nikitaved) [reply aten/src/ATen/native/cuda/SparseBinaryOpIntersectionKernel.cu:88]: No allocation in TI, everything is not coalesced. I had everything restricted, but thought maybe const-ness might have been sufficient...

**Review by amjames — APPROVED** [review]
This is a great improvement! Thanks

**Comment by nikitaved** [comment]
@cpuhrsch , do you have any suggestions/comments?

**Review by cpuhrsch — APPROVED** [review]
Great! Thanks!

**Comment by cpuhrsch** [comment]
@pytorch merge

**Pull request closed without merging**